File: frontend/portal/src/bezugsArt.test.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { describe, expect, it } from 'vitest';
import {
  ALLE,
  EINHEIT_DER_MESSSTELLE,
  einheitenDer,
  geltungDer,
  herkunftDer,
  periodenDer,
  pruefen,
  type Art,
  type ArtEingang,
  type Vokabular,
} from './bezugsArt';

/**
 * Das Modul der ARTEN (UEMS AP-09 §4.2) gegen den Block `arten` der geteilten
 * Vektor-Datei `docs/contracts/v2/bezugsdaten-vectors.json`. Der Java-Zwilling
 * `BezugsArtTest` fährt dieselbe Datei — beide grün heißt: dieselbe Antwort auf
 * jede Prüfung. Die Schema-Treue des Blocks prüft `bezugsdaten.test.ts`.
 */

type Json = any;

const V2 = resolve(process.cwd(), '../../docs/contracts/v2');
const lies = (datei: string): Json => JSON.parse(readFileSync(resolve(V2, datei), 'utf8'));
const vectors: Json = lies('bezugsdaten-vectors.json');
const referenz: Json = lies('uems-referenzunternehmen.json');

const arten = vectors.arten.je_art as Record<string, Art>;
const pruefungen = vectors.arten.pruefungen as Array<{
  name: string;
  art: string;
  beispiel: string | null;
  eingang: ArtEingang;
  ergebnis: { abweichend: string[] };
  hinweis?: string;
}>;
const vok: Vokabular = {
  wertart: vectors.vokabulare.wertart,
  geltung_art: vectors.vokabulare.geltung_art,
  periode_art: vectors.vokabulare.periode_art,
  herkunft_art: vectors.vokabulare.herkunft_art,
  einheiten: vectors.einheiten,
};
const alleEinheiten = Object.values(vok.einheiten).flat();

describe('bezugsArt — jede Art spricht nur Wörter der vorhandenen Vokabulare', () => {
  it('das Vokabular der Arten ist nicht leer', () => {
    expect(Object.keys(arten).length).toBeGreaterThan(0);
  });

  for (const [wort, art] of Object.entries(arten)) {
    it(`${wort} · ${art.name}`, () => {
      expect(art.name.trim()).not.toBe('');
      if (Array.isArray(art.einheiten)) {
        expect(art.einheiten.length).toBeGreaterThan(0);
        expect(new Set(art.einheiten).size).toBe(art.einheiten.length);
        // Die Einheiten fragt das Modul bei bezugsEinheit an — keine fällt heraus.
        expect(einheitenDer(art, vok.einheiten, null)).toEqual(art.einheiten);
      } else {
        expect([ALLE, EINHEIT_DER_MESSSTELLE]).toContain(art.einheiten);
      }
      expect(vok.wertart).toContain(art.wertart);
      expect(new Set(art.perioden).size).toBe(art.perioden.length);
      expect(periodenDer(art, vok.periode_art)).toEqual(art.perioden);
      // Nur ein Periodenwert hat Perioden (M1).
      expect(art.perioden.length === 0).toBe(art.wertart !== 'periodenwert');
      if (Array.isArray(art.geltung)) {
        expect(art.geltung.length).toBeGreaterThan(0);
        expect(new Set(art.geltung).size).toBe(art.geltung.length);
        expect(geltungDer(art, vok.geltung_art)).toEqual(art.geltung);
      } else {
        expect(art.geltung).toBe(ALLE);
      }
      expect(art.herkunft.length).toBeGreaterThan(0);
      expect(new Set(art.herkunft).size).toBe(art.herkunft.length);
      expect(herkunftDer(art, vok.herkunft_art)).toEqual(art.herkunft);
    });
  }
});

describe('bezugsArt — die Prüfungen der Datei', () => {
  for (const p of pruefungen) {
    it(p.name, () => {
      expect(pruefen(p.art, p.eingang, arten, vok)).toEqual(p.ergebnis.abweichend);
    });
  }
});

describe('bezugsArt — die Zusagen ohne eigene Prüfung', () => {
  it('die Beispiele sind GENAU die Bezugsgrößen des Referenzunternehmens und passen zu ihrer Art', () => {
    const je = new Map<string, Json>((referenz.bezugsgroessen as Json[]).map((b) => [b.kennzeichen, b]));
    const beispiele = new Set<string>();
    for (const p of pruefungen) {
      if (p.beispiel === null) continue;
      beispiele.add(p.beispiel);
      const b = je.get(p.beispiel);
      expect(b, `${p.beispiel} im Referenzunternehmen`).toBeDefined();
      expect(p.ergebnis.abweichend, `${p.beispiel} passt zu seiner Art`).toEqual([]);
      expect(p.eingang.einheit).toBe(b.einheit_code);
      expect(p.eingang.periode_art).toBe(b.periode_code);
      expect(p.eingang.wertart).toBe(b.wertart);
      if (b.geltung_art !== p.eingang.geltung_art) {
        expect(p.hinweis ?? '', `${p.beispiel}: die Abweichung nennt ihren Grund`).toContain(`„${b.geltung_art}“`);
      }
    }
    expect([...beispiele].sort()).toEqual([...je.keys()].sort());
  });

  it('eine Einheit, die die Art nicht führt, passt nicht — auch wenn sie im Vokabular steht', () => {
    let abgelehnt = 0;
    for (const [wort, art] of Object.entries(arten)) {
      const messstelle = art.einheiten === EINHEIT_DER_MESSSTELLE ? alleEinheiten[0] : null;
      const gefuehrt = einheitenDer(art, vok.einheiten, messstelle);
      expect(gefuehrt.length, `${wort} führt Einheiten`).toBeGreaterThan(0);
      for (const einheit of alleEinheiten) {
        const eingang: ArtEingang = {
          wertart: art.wertart,
          geltung_art: geltungDer(art, vok.geltung_art)[0],
          einheit,
          periode_art: art.perioden[0] ?? null,
          herkunft_art: art.herkunft[0],
          einheit_der_messstelle: messstelle,
        };
        const soll = gefuehrt.includes(einheit) ? [] : ['einheit'];
        expect(pruefen(wort, eingang, arten, vok), `${wort} in ${einheit}`).toEqual(soll);
        abgelehnt += soll.length;
      }
    }
    expect(abgelehnt).toBeGreaterThan(0);
  });

  it('keine zweite Liste: das Modul nennt keine Art und kein Wort eines Vokabulars', () => {
    const quelle = readFileSync(resolve(process.cwd(), 'src/bezugsArt.ts'), 'utf8');
    const woerter = new Set<string>([
      ...Object.keys(arten),
      ...Object.values(arten).map((a) => a.name),
      ...alleEinheiten,
      ...vok.wertart,
      ...vok.geltung_art,
      ...vok.periode_art,
      ...vok.herkunft_art,
    ]);
    // Die beiden Verweise der Datei sind Wörter des Moduls; „messstelle" ist zugleich ein Geltungsbereich.
    woerter.delete(ALLE);
    woerter.delete(EINHEIT_DER_MESSSTELLE);
    for (const wort of woerter) {
      expect(quelle.includes(`'${wort}'`) || quelle.includes(`"${wort}"`), `bezugsArt.ts nennt „${wort}"`).toBe(false);
    }
  });
});

File: services/api/src/test/java/com/voltpilot/api/uems/BezugsArtTest.java
package com.voltpilot.api.uems;

import static org.assertj.core.api.Assertions.assertThat;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.voltpilot.api.uems.BezugsArt.Art;
import com.voltpilot.api.uems.BezugsArt.Auswahl;
import com.voltpilot.api.uems.BezugsArt.Eingang;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

/**
 * Das Modul der ARTEN (UEMS AP-09 §4.2) gegen den Block {@code arten} der geteilten Vektor-Datei
 * {@code docs/contracts/v2/bezugsdaten-vectors.json}. Der TS-Zwilling {@code bezugsArt.test.ts}
 * fährt dieselbe Datei — beide grün heißt: dieselbe Antwort auf jede Prüfung.
 *
 * <p>Die Schema-Treue des Blocks prüft {@code BezugsdatenVectorsTest.dieDateiHaeltIhrSchema}.
 */
class BezugsArtTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static final Path V2 = Path.of("..", "..", "docs", "contracts", "v2");

    private static final Path MODUL =
            Path.of("src", "main", "java", "com", "voltpilot", "api", "uems", "BezugsArt.java");

    private static JsonNode lies(Path pfad) throws Exception {
        return MAPPER.readTree(Files.readString(pfad));
    }

    private static JsonNode vektoren() throws Exception {
        return lies(V2.resolve("bezugsdaten-vectors.json"));
    }

    private static List<String> texte(JsonNode liste) {
        List<String> aus = new ArrayList<>();
        liste.forEach(w -> aus.add(w.asText()));
        return List.copyOf(aus);
    }

    private static Auswahl auswahl(JsonNode knoten) {
        return knoten.isArray() ? Auswahl.liste(texte(knoten)) : Auswahl.verweis(knoten.asText());
    }

    private static List<Art> arten(JsonNode v) {
        List<Art> alle = new ArrayList<>();
        v.path("arten").path("je_art").properties().forEach(e -> {
            JsonNode a = e.getValue();
            alle.add(new Art(
                    e.getKey(),
                    a.path("name").asText(),
                    auswahl(a.path("einheiten")),
                    a.path("wertart").asText(),
                    texte(a.path("perioden")),
                    auswahl(a.path("geltung")),
                    texte(a.path("herkunft"))));
        });
        return List.copyOf(alle);
    }

    private static BezugsgroesseRegeln.Vokabular vokabular(JsonNode v) {
        Map<String, List<String>> einheiten = new LinkedHashMap<>();
        v.path("einheiten").properties().forEach(e -> einheiten.put(e.getKey(), texte(e.getValue())));
        return new BezugsgroesseRegeln.Vokabular(
                texte(v.at("/vokabulare/wertart")),
                texte(v.at("/vokabulare/geltung_art")),
                texte(v.at("/vokabulare/periode_art")),
                einheiten);
    }

    private static String text(JsonNode n) {
        return n.isNull() || n.isMissingNode() ? null : n.asText();
    }

    private static Eingang eingang(JsonNode e) {
        return new Eingang(
                text(e.path("wertart")),
                text(e.path("geltung_art")),
                text(e.path("einheit")),
                text(e.path("periode_art")),
                text(e.path("herkunft_art")),
                text(e.path("einheit_der_messstelle")));
    }

    /** Je Art: jede Angabe ist ein Wort der VORHANDENEN Vokabulare — die Einheiten fragt das Modul bei {@link BezugsEinheit} an. */
    @TestFactory
    List<DynamicTest> jedeArtSprichtNurWoerterDerVorhandenenVokabulare() throws Exception {
        JsonNode v = vektoren();
        BezugsgroesseRegeln.Vokabular vok = vokabular(v);
        List<String> herkunftArten = texte(v.at("/vokabulare/herkunft_art"));
        List<DynamicTest> tests = new ArrayList<>();
        for (Art art : arten(v)) {
            tests.add(DynamicTest.dynamicTest(art.art() + " · " + art.name(), () -> {
                assertThat(art.name()).as("Kundenwort").isNotBlank();
                if (art.einheiten().aus() == null) {
                    assertThat(art.einheiten().woerter()).as("Einheiten").isNotEmpty().doesNotHaveDuplicates();
                    assertThat(BezugsArt.einheitenDer(art, vok.einheiten(), null))
                            .as("jede Einheit steht im Einheiten-Vokabular")
                            .containsExactlyElementsOf(art.einheiten().woerter());
                } else {
                    assertThat(art.einheiten().aus())
                            .as("Verweis der Einheiten")
                            .isIn(BezugsArt.ALLE, BezugsArt.EINHEIT_DER_MESSSTELLE);
                }
                assertThat(vok.wertarten()).as("Wertart").contains(art.wertart());
                assertThat(art.perioden()).doesNotHaveDuplicates();
                assertThat(BezugsArt.periodenDer(art, vok.periodeArten()))
                        .as("jede Periode steht in vokabulare.periode_art")
                        .containsExactlyElementsOf(art.perioden());
                assertThat(art.perioden().isEmpty())
                        .as("nur ein Periodenwert hat Perioden (M1)")
                        .isEqualTo(!BezugsgroesseRegeln.PERIODENWERT.equals(art.wertart()));
                if (art.geltung().aus() == null) {
                    assertThat(art.geltung().woerter()).isNotEmpty().doesNotHaveDuplicates();
                    assertThat(BezugsArt.geltungDer(art, vok.geltungArten()))
                            .as("jeder Geltungsbereich steht in vokabulare.geltung_art")
                            .containsExactlyElementsOf(art.geltung().woerter());
                } else {
                    assertThat(art.geltung().aus()).isEqualTo(BezugsArt.ALLE);
                }
                assertThat(art.herkunft()).isNotEmpty().doesNotHaveDuplicates();
                assertThat(BezugsArt.herkunftDer(art, herkunftArten))
                        .as("jede Herkunft steht in vokabulare.herkunft_art")
                        .containsExactlyElementsOf(art.herkunft());
            }));
        }
        assertThat(tests).as("das Vokabular der Arten ist nicht leer").isNotEmpty();
        return tests;
    }

    /** Jede Prüfung des Blocks — dieselbe Antwort, die der TS-Zwilling gibt. */
    @TestFactory
    List<DynamicTest> diePruefungenDerDatei() throws Exception {
        JsonNode v = vektoren();
        List<Art> arten = arten(v);
        BezugsgroesseRegeln.Vokabular vok = vokabular(v);
        List<String> herkunftArten = texte(v.at("/vokabulare/herkunft_art"));
        List<DynamicTest> tests = new ArrayList<>();
        for (JsonNode p : v.at("/arten/pruefungen")) {
            tests.add(DynamicTest.dynamicTest(p.path("name").asText(), () -> assertThat(BezugsArt.pruefen(
                            p.path("art").asText(), eingang(p.path("eingang")), arten, vok, herkunftArten))
                    .containsExactlyElementsOf(texte(p.at("/ergebnis/abweichend")))));
        }
        assertThat(tests).isNotEmpty();
        return tests;
    }

    /**
     * Die Beispiele sind GENAU die Bezugsgrößen des Referenzunternehmens, jede passt zu ihrer Art,
     * und ihr Eingang ist der des Referenzunternehmens — eine Abweichung nennt ihren Grund.
     */
    @Test
    void dieBeispieleSindDieBezugsgroessenDesReferenzunternehmensUndPassen() throws Exception {
        JsonNode v = vektoren();
        Map<String, JsonNode> referenz = new LinkedHashMap<>();
        lies(V2.resolve("uems-referenzunternehmen.json"))
                .path("bezugsgroessen")
                .forEach(b -> referenz.put(b.path("kennzeichen").asText(), b));
        Set<String> beispiele = new LinkedHashSet<>();
        for (JsonNode p : v.at("/arten/pruefungen")) {
            String kz = text(p.path("beispiel"));
            if (kz == null) {
                continue;
            }
            beispiele.add(kz);
            JsonNode b = referenz.get(kz);
            JsonNode ein = p.path("eingang");
            assertThat(b).as(kz + " im Referenzunternehmen").isNotNull();
            assertThat(texte(p.at("/ergebnis/abweichend"))).as(kz + " passt zu seiner Art").isEmpty();
            assertThat(text(ein.path("einheit"))).as(kz + " · einheit").isEqualTo(text(b.path("einheit_code")));
            assertThat(text(ein.path("periode_art"))).as(kz + " · periode").isEqualTo(text(b.path("periode_code")));
            assertThat(text(ein.path("wertart"))).as(kz + " · wertart").isEqualTo(text(b.path("wertart")));
            String geltung = text(b.path("geltung_art"));
            if (!geltung.equals(text(ein.path("geltung_art")))) {
                assertThat(p.path("hinweis").asText())
                        .as(kz + ": die Abweichung vom Referenzunternehmen nennt ihren Grund")
                        .contains("„" + geltung + "“");
            }
        }
        assertThat(beispiele).as("BZ-1 … BZ-7").containsExactlyInAnyOrderElementsOf(referenz.keySet());
    }

    /** Eine Einheit, die die Art nicht führt, passt nicht — auch wenn sie im Vokabular steht. */
    @Test
    void eineEinheitDieDieArtNichtFuehrtWirdAbgelehnt() throws Exception {
        JsonNode v = vektoren();
        List<Art> arten = arten(v);
        BezugsgroesseRegeln.Vokabular vok = vokabular(v);
        List<String> herkunftArten = texte(v.at("/vokabulare/herkunft_art"));
        List<String> alleEinheiten = new ArrayList<>();
        vok.einheiten().values().forEach(alleEinheiten::addAll);
        int abgelehnt = 0;
        for (Art art : arten) {
            String messstelle = BezugsArt.EINHEIT_DER_MESSSTELLE.equals(art.einheiten().aus()) ? alleEinheiten.get(0) : null;
            List<String> gefuehrt = BezugsArt.einheitenDer(art, vok.einheiten(), messstelle);
            assertThat(gefuehrt).as(art.art() + " führt Einheiten").isNotEmpty();
            String periode = art.perioden().isEmpty() ? null : art.perioden().get(0);
            String geltung = BezugsArt.geltungDer(art, vok.geltungArten()).get(0);
            for (String einheit : alleEinheiten) {
                Eingang e = new Eingang(art.wertart(), geltung, einheit, periode, art.herkunft().get(0), messstelle);
                List<String> soll = gefuehrt.contains(einheit) ? List.of() : List.of("einheit");
                assertThat(BezugsArt.pruefen(art.art(), e, arten, vok, herkunftArten))
                        .as(art.art() + " in " + einheit)
                        .containsExactlyElementsOf(soll);
                abgelehnt += soll.size();
            }
        }
        assertThat(abgelehnt).as("es gibt Einheiten, die eine Art nicht führt").isPositive();
    }

    /** Keine zweite Liste: das Modul nennt keine Art und kein Wort eines Vokabulars — alles kommt herein. */
    @Test
    void dasModulFuehrtKeineZweiteListe() throws Exception {
        JsonNode v = vektoren();
        String quelle = Files.readString(MODUL);
        Set<String> woerter = new LinkedHashSet<>();
        for (Art art : arten(v)) {
            woerter.add(art.art());
            woerter.add(art.name());
        }
        v.path("einheiten").forEach(g -> woerter.addAll(texte(g)));
        for (String liste : List.of("wertart", "geltung_art", "periode_art", "herkunft_art")) {
            woerter.addAll(texte(v.path("vokabulare").path(liste)));
        }
        // Die beiden Verweise der Datei sind Wörter des Moduls; „messstelle“ ist zugleich ein Geltungsbereich.
        woerter.removeAll(List.of(BezugsArt.ALLE, BezugsArt.EINHEIT_DER_MESSSTELLE));
        for (String wort : woerter) {
            assertThat(quelle).as("BezugsArt.java nennt „" + wort + "“").doesNotContain("\"" + wort + "\"");
        }
    }
}
